src/main/java/edu/harvard/iq/dataverse/api/Metrics.java

```diff
@@ -26,404 +41,836 @@
 
-        } catch (Exception ex) {
-            return error(BAD_REQUEST, ex.getLocalizedMessage());
+        if (null == jsonArray) { // run query and save
+            // Only handling published right now
+            jsonArray = metricsSvc.fileDownloadsTimeSeries(d, true);
+            metricsSvc.save(new Metric(metricName, null, null, d, jsonArray.toString()));
         }
+        MediaType requestedType = getVariant(req, MediaType.valueOf(FileUtil.MIME_TYPE_CSV), MediaType.APPLICATION_JSON_TYPE);
+        if ((requestedType != null) && (requestedType.equals(MediaType.APPLICATION_JSON_TYPE))) {
+            return ok(jsonArray);
+        }
+        return ok(FileUtil.jsonToCSV(jsonArray, MetricsUtil.DATE, MetricsUtil.ID, MetricsUtil.PID, MetricsUtil.COUNT), MediaType.valueOf(FileUtil.MIME_TYPE_CSV), "uniquefiledownloads.timeseries.csv");
+    }
+    
+    @GET
+    @Path("tree")
+    public Response getDataversesTree(@Context UriInfo uriInfo, @QueryParam("parentAlias") String parentAlias) {
+        return getDataversesTreeToMonth(uriInfo, MetricsUtil.getCurrentMonth(), parentAlias);
+    }
+
+    @GET
+    @Path("tree/toMonth/{yyyymm}")
+    public Response getDataversesTreeToMonth(@Context UriInfo uriInfo, @PathParam("yyyymm") String yyyymm, @QueryParam("parentAlias") String parentAlias) {
+
+        Dataverse d = findDataverseOrDieIfNotFound(parentAlias);
+
+        try {
+            errorIfUnrecongizedQueryParamPassed(uriInfo, new String[] { "parentAlias" });
+        } catch (IllegalArgumentException ia) {
+            return error(BAD_REQUEST, ia.getLocalizedMessage());
+        }
+
+        String metricName = "tree";
+        String sanitizedyyyymm = MetricsUtil.sanitizeYearMonthUserInput(yyyymm);
+
+        JsonObject jsonObj = MetricsUtil.stringToJsonObject(metricsSvc.returnUnexpiredCacheMonthly(metricName, sanitizedyyyymm, null, d));
+        if (null == jsonObj) { // run query and save
+            jsonObj = metricsSvc.getDataverseTree(d, sanitizedyyyymm, DatasetVersion.VersionState.RELEASED);
+            metricsSvc.save(new Metric(metricName, sanitizedyyyymm, null, d, jsonObj.toString()));
+        }
+        return ok(jsonObj);
     }
 
     private void errorIfUnrecongizedQueryParamPassed(UriInfo uriDetails, String[] allowedQueryParams) throws IllegalArgumentException {
-        for(String theKey : uriDetails.getQueryParameters().keySet()) {
-            if(!Arrays.stream(allowedQueryParams).anyMatch(theKey::equals)) {
+        for (String theKey : uriDetails.getQueryParameters().keySet()) {
+            if (!Arrays.stream(allowedQueryParams).anyMatch(theKey::equals)) {
                 throw new IllegalArgumentException("queryParameter " + theKey + " not supported for this endpont");
             }
         }
-        
+
     }
-    
+
+    // Throws a WebApplicationException if alias is not null and Dataverse can't be
+    // found
+    private Dataverse findDataverseOrDieIfNotFound(String alias) {
```

Comment:
I'm used to the "orDie" methods working including the original findDataverseOrDie to work in certain way, to work on unpublished things, to throw a WrappedResponse when it can't be found. Since this method only returns published dataverses maybe it could be renamed? This is just a passing thought. I can also live with it with the current name.

Reply:
Not sure I understand why this doesn't look like an orDie method - if the alias requested doesn't exist, it throws an exception (a WebApplicationException rather than WrappedResponse to use the updated error handling that's been added). Is it what is returned that is an issue or something else? Or just findPublishedDataverseOrDie ?